Swift code guidelines.md

```diff
@@ -0,0 +1,196 @@
+```swift
+if name == "okta" { ... }
+switch tokenType { ... }
+let formattedTokens = tokens.map { $0 + "\\" }
+tokens.forEach { number in print(number) }
+```
+**Not preferred:**
+```swift
+if (name == "okta") { ... }
+switch (tokenType) { ... }
+let formattedTokens = tokens.map() { $0 + "\\" }
+tokens.forEach { (number) in print(number) }
+```
+3.8. Use compiler inferred context.
+
+**Preferred:**
+```swift
+view.backgroundColor = .red
+let customView = UIView(frame: .zero)
+let selector = #selector(viewDidLoad)
+let message = "Hello Okta!"
+```
+**Not preferred:**
+```swift
+view.backgroundColor = UIColor.red
+let customView = UIView(frame: CGRect.zero)
+let selector = #selector(MyViewController.viewDidLoad)
+let message: String = "Hello Okta!"
+```
+3.9. Name unused closure parameters as underscores `_`.
+3.10. Avoid nested `if/else` statements. Use `Happy Path` rule. 
+
+**Preferred:**
+```swift
+func message(from response: Response) -> String? {
+  guard let context = response.context else {
+    return nil
+  }
+
+  guard context.isAuthenticated else {
+    logOut()
+    return nil
+  }
+
+  return "User is authenticated"
+}
+```
+**Not preferred:**
+```swift
+func message(from response: Response) -> String? {
+  if let context = response.context {
+      if context.isAuthenticated {
+        return "User is authenticated" 
+      } else {
+        logOut()
+        return nil 
+      }
+  } else {
+    return nil
+  }
+}
+```
+
+3.11. Use trailing closure syntax when **single** closure parameter.
+**Preferred:**
+```swift
+sendRequest(request) { response, error in
+ ...
+}
+```
+**Not Preferred:**
+```swift
+sendRequest(request, completion: { response, error in
+ ...
+})
+```
+
+3.12. Use multi-line string literal.
+
+**Preferred:**
+```swift
+let testOktaToken = """
+    Lorem Ipsum is simply dummy text of the printing and typesetting industry. \
+    Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, \
+    when an unknown printer took a galley of type and scrambled it to make \
+    a type specimen book.
+    """
+```
+
+3.13. Use multi-line `guard` statements.
+
+**Preferred:**
+```swift
+guard let user = response["user"] as? User,
+      let username = user.username
+else {
+    return
+}
+```
+**Not preferred:**
+```swift
+guard let user = response["user"] as? User, let username = user.username else {
+    return
+}
+
+guard let user = response["user"] as? User,
+      let username = user.username else {
+    return
+}
+```
+
+3.14. Use [XCTUnwrap](https://developer.apple.com/documentation/xctest/3380195-xctunwrap) instead of forced unwrapping in tests.
+
+#### 4. Don'ts
+
+4.1. Don't use `fatalError` or `precondition`. Prefer `assert` if it is recoverable case. If not, handle it in proper way. 
+4.2. Avoid implicit unwrapping unless you're 100% sure that a value exists. Implicit unwrapping causes crash if a value is `nil`. Therefore, handle unsuccess paths in code (e.g. logging, return an error etc.).  
```

Comment:
I would make this stronger. 
```suggestion
4.2. Avoid implicit unwrapping (`!`), as this will cause a crash if a value is `nil`. Unwrap using `guard let` or `if let` and handle error cases in code as appropriate (e.g. logging, return an error etc.).  
```